src/parser/mod.rs

```diff
@@ -5199,13 +5199,20 @@ impl<'a> Parser<'a> {
 
         // parse: [ argname ] argtype
         let mut name = None;
+        let next_token = self.peek_token();
```

Comment:
> The code you proposed does not work as Int2 (or any analogous such type) does not fall in if let DataType::Custom(n, _) = &data_type {

Oh how did you mean here by `Int2` in this example not being parsed as a custom datatype, do we get back a different type or does `parse_data_type` fail in that scenario?

I think ideally we will want to do without this `self.peek_token()` to avoid the cloning that it includes

Reply:
The argument named `Int2` (as described in the [issue](https://github.com/apache/datafusion-sqlparser-rs/issues/1825)) is not parsed as `DataType::Custom`, but as a [`DataType::Int2`](https://github.com/apache/datafusion-sqlparser-rs/blob/a464f8e8d7a5057e4e5b8046b0f619acdf7fce74/src/ast/data_type.rs#L150). Analogously, any other such argument names that collides with data types from other SQL engines would be parsed into a type.

Now, if I were to convert back to string `DataType::Int2` I would get some arbitrary capitalization which in this case is `INT2` - without the `peek_token`, I am unsure how we can preserve the initial token from being lost.

Reply:
Ah I see that makes sense! Maybe something like this we can do to restrict the cloning to only when necessary?
```rust
let data_type_idx = self.get_current_index();
if let Some(next_data_type) = self.maybe_parse(|parser| {
    name = parser.token_at(data_type_idx).to_string();
   // ...
})
```
Coming to think about it, would we not need to sanity check that the first token is actually a `Token::Word` variant? current code seems to assume that to be the case which might not necessarily be true.
For example following how the following sql would be parsed, we can probably have a test case it
```rust
function(struct<a,b> int64)
```
we would call `to_string()` on only the first token which would be `struct` even though this query is technically invalid?

Reply:
Could you provide a complete example of such a broken case, so that I may add it to the test suite?

Reply:
I think something like this potentially
```rust
let sql = r#"CREATE OR REPLACE FUNCTION foo(a TIMESTAMP WITH TIMEZONE, b VARCHAR) RETURNS BOOLEAN LANGUAGE plpgsql AS $$
BEGIN
	RETURN TRUE;
END;
$$"#;
pg().verified_stmt(sql);
```